pkg/apis/core/validation/validation.go

```diff
@@ -3976,6 +3978,20 @@ func ValidateTolerations(tolerations []core.Toleration, fldPath *field.Path) fie
 		if len(toleration.Effect) > 0 {
 			allErrors = append(allErrors, validateTaintEffect(&toleration.Effect, true, idxPath.Child("effect"))...)
 		}
+
+		// validate if taint is unique by <key, effect>
+		if len(uniqueTaints[toleration.Effect]) > 0 && uniqueTaints[toleration.Effect].Has(toleration.Key) {
+			duplicatedError := field.Duplicate(idxPath, toleration)
+			duplicatedError.Detail = "taints must be unique by key and effect pair"
+			allErrors = append(allErrors, duplicatedError)
```

Comment:
I'm afraid this will be a breaking change for the to be created Pods/PodTemplates but also for the existing ones.
To be checked with the maintainers, but maybe all we can do is deduplicate the tolerations before storing and make sure duplication doesn't break equality...

If we really want to enforce this for new objects, we'll need to migrate/deduplicate the existing ones in the store (maybe sth like https://kubernetes.io/docs/tasks/manage-kubernetes-objects/storage-version-migration/ could help) and then we can require this.

Reply:
duplicate toleration error will be prompted when the existing one with  duplicate toleration is updated,  but I think it is acceptable, because  an update operation occurs, the pod will be rebuilt.  At this time, duplicate toleration in pod template is removed.

Reply:
i aslo add `UniqueToleration` feature gate